### ISSUE TITLE
Add Phantasialand Ride Times 

### ISSUE THREAD

**Issue by WolfDevEF** [cubehouse/themeparks#104]
The new app from Phantasialand now offers the waiting times.
https://play.google.com/store/apps/details?id=de.phantasialand

**Comment by tpxtron** [comment]
I already spent some time on reverse engineering how the app works. I would like to share what I've got  so far, maybe somebody finds that information useful and/or would like to help me getting it done:

The main API endpoint is `https://api.phlsys.de/api/``

### POIs
URL: `https://api.phlsys.de/api/pois`
You get a list of all POIs ( = rides/shows/restaurants/etc.) in JSON format.

Example entry:
```JSON
{"poiNumber":100,"poiNumberWinter":null,"minAge":null,"maxAge":null,"minSize":100,"maxSize":null,"minSizeEscort":0,"tags":["ATTRACTION_TYPE_CHILDREN","ATTRACTION_TYPE_MIND_THE_SIZE_CHART"],"category":"ATTRACTIONS","slug":"Bolles Flugschule","slugWinter":null,"area":"BERLIN","parkMonitorReferenceName":"Bolles Flugschule","seasons":["SUMMER","WINTER"],"navigationEnabled":true,"navigationEnabledWinter":true,"attachedToNavigationGraph":false,"attachedToNavigationGraphWinter":true,"adminOnly":false,"id":1,"titleImageId":2,"titleImageWinterId":null,"createdAt":"2018-03-14T15:45:47.000Z","updatedAt":"2018-11-27T15:44:23.000Z","_title":"Bolles Flugschule","_titleWinter":null,"_tagline":"Interaktives Kinderkarussell in Berlin","_taglineWinter":null,"_description":"Bei Bolles Flugschule geht es für die Kinder munter im Kreis durch die Berliner Luft. Ob im Doppeldecker oder im Zeppelin – hier geht es hoch hinaus und rund herum!","_descriptionWinter":null,"_entrance":{"id":null,"world":{"lat":50.79900942373939,"lng":6.879732012748718},"map":{"lat":50.79930567134277,"lng":6.879249215126038}},"_entranceWinter":{"id":null,"world":{"lat":50.79900942373939,"lng":6.879732012748718},"map":{"lat":50.79930567134277,"lng":6.879249215126038}},"_exit":null,"_exitWinter":null,"_preferredDestinations":[],"_preferredDestinationsWinter":[],"titleImage":{"id":2,"url":"https://live-phlsys.s3.amazonaws.com/6a4805a062d8e29dbb2d46b236a40e7f.jpg"}}
```
The `number` property corresponds to the number in the printed park map. Also, you can get information on a single POI by appending the index (starting at "1") from the acquired array to the URL: `https://api.phlsys.de/api/pois/1`

This returns the following JSON data:
```JSON
{"poiNumber":100,"poiNumberWinter":null,"minAge":null,"maxAge":null,"minSize":100,"maxSize":null,"minSizeEscort":0,"tags":["ATTRACTION_TYPE_CHILDREN","ATTRACTION_TYPE_MIND_THE_SIZE_CHART"],"category":"ATTRACTIONS","slug":"Bolles Flugschule","slugWinter":null,"area":"BERLIN","parkMonitorReferenceName":"Bolles Flugschule","seasons":["SUMMER","WINTER"],"navigationEnabled":true,"navigationEnabledWinter":true,"attachedToNavigationGraph":false,"attachedToNavigationGraphWinter":true,"adminOnly":false,"id":1,"titleImageId":2,"titleImageWinterId":null,"createdAt":"2018-03-14T15:45:47.000Z","updatedAt":"2018-11-27T15:44:23.000Z","_title":"Bolles Flugschule","_titleWinter":null,"_tagline":"Interaktives Kinderkarussell in Berlin","_taglineWinter":null,"_description":"Bei Bolles Flugschule geht es für die Kinder munter im Kreis durch die Berliner Luft. Ob im Doppeldecker oder im Zeppelin – hier geht es hoch hinaus und rund herum!","_descriptionWinter":null,"_entrance":{"id":null,"world":{"lat":50.79900942373939,"lng":6.879732012748718},"map":{"lat":50.79930567134277,"lng":6.879249215126038}},"_entranceWinter":{"id":null,"world":{"lat":50.79900942373939,"lng":6.879732012748718},"map":{"lat":50.79930567134277,"lng":6.879249215126038}},"_exit":null,"_exitWinter":null,"_preferredDestinations":[],"_preferredDestinationsWinter":[],"titleImage":{"id":2,"url":"https://live-phlsys.s3.amazonaws.com/6a4805a062d8e29dbb2d46b236a40e7f.jpg"}}
```

### Waiting/Show times
URL: `https://api.phlsys.de/api/signage-snapshots`
Also, there is an endpoint for waiting and show times. This endpoint, however, requires authentication and some more details.

I managed to reverse-engineer the GET parameters that need to be appended as follows:

```JAVA
public static void main() {
    // These are coordinates from somewhere in the park
    Double lat = 50.799449980117394;
    Double lng = 6.880099140107632;

    long doubleToLongBits = Double.doubleToLongBits(lat);
    int i = ((int) (doubleToLongBits ^ (doubleToLongBits >>> 32))) * 31;
    long doubleToLongBits2 = Double.doubleToLongBits(lng);
    // Append this as 'loc' GET parameter to the request:
    System.out.println(i + ((int) ((doubleToLongBits2 >>> 32) ^ doubleToLongBits2)));

    // The Beacon IDs need to be obtained from within the park.
    // There are Bluetooth beacons hidden all across the park supporting GPS positioning (e.g. inside buildings
    // You need to use a bluetooth beacon scanner app to get beacon IDs from a spot inside the park and write down the GPS coordinates
    List<String> beaconIds = new ArrayList<String>();
    beaconIds.add("ID 1");
    beaconIds.add("ID 2");
    // Append this as 'tokens' GET parameter to the request:
    System.out.println(beaconIds.hashCode());
  }
```

An example Request should look as follows: `https://api.phlsys.de/api/signage-snapshots?loc=1485692860&tokens=E1A1D048-D592-A289-0742-E2B7952C69EA`

However, you need to have some kind of authentication, otherwise the request returns a 401 response.

This is where I stopped researching as I ran out of time to continue.

There is some kind of OAuth-Style token that needs to be generated. When the app is launched for the first time, it registers itself via the API and stores a randomized (?) username/password in the device's secure storage. It then gets an API token from somewhere and uses the token in the header for authentication. The username/password is later needed to get a new token when the old one expires.

In case anyone would like to help me, feel free to contact me. I'll be glad to share any further knowledge and hopefully we can get this up and running some time in the future.

**Comment by rambleraptor** [comment]
I'd recommend trying to run the app through a man-in-the-middle proxy. I personally recommend [mitmproxy](https://mitmproxy.org/).

Once set up, the proxy will capture all of the HTTP requests as they happen. It's a lot easier to understand what tokens are being sent and where they come from.

**Comment by tpxtron** [comment]
The app does certificate pinning. So, MITM won‘t be possible.

**Comment by rambleraptor** [comment]
Huh, that's very unusual. I haven't heard of a theme park app that uses its own certs before.

**Comment by cubehouse** [comment]
Rather than using their certs, certificate pinning embeds some sort of identifier of the expected certificate within the app (a hash or something). When making the HTTPS request, if the provided certificate doesn't match the embedded identifier, the connection is rejected.

There are ways around this, if you search for "Certificate Bypassing Android Frida" or similar. Phantasialand and Efteling are the only two park apps I am aware of that require certificate pinning bypassing though (so far), and it's relatively easy to setup. It just means you need to fork out for a long-lasting SSL certificate, or you will need to push an update for your app out very regularly with new certificate identifiers. I don't think either of these are worth the effort, but some people clearly do.

The issue with Phantasialand specifically though is that they also have Estimote Bluetooth Beacons around the park. Your phone listens for these and sends the beacon IDs to the server when requesting wait times. This means you need to be in the park to get those IDs correctly to get wait times. It's pretty effective at stopping GPS spoofing! Again, I don't see how it's worth the effort, but there we go.

**Comment by BenediktCleff** [comment]
They‘ve removed the need for a Bluetooth connection in the lastest version of the Phantasialand app. Maybe it’s worth the effort, to recheck their current authorization. I hope I‘ll get to do it in the coming days...

**Comment by cubehouse** [comment]
Initial work done in #196 

Need to finish up opening times.

**Comment by cubehouse** [comment]
Published with 5.1.9.

**Comment by DieserMerlin** [comment]
This does not work anymore. Saying "AUTHORIZATION_REQUIRED" on every request.
Any idea how to fix?
